### PR TITLE
3970.type annotations for wormholetesting

### PR DESCRIPTION
**Pull request by danielzwlee**
Type annotations added to wormholetesting.py.

### CODE REVIEW

File: src/allmydata/test/cli/wormholetesting.py [hunk 1]
```diff
@@ -61,23 +60,23 @@ class MemoryWormholeServer(object):
     :ivar _waiters: Observers waiting for a wormhole to be created for a
         specific application id and relay URL combination.
     """
-    _apps: dict[ApplicationKey, _WormholeApp] = field(default=Factory(dict))
-    _waiters: dict[ApplicationKey, Deferred] = field(default=Factory(dict))
+    _apps: Dict[ApplicationKey, _WormholeApp] = field(default=Factory(dict))
```
Comment (itamarst) [inline src/allmydata/test/cli/wormholetesting.py:63]: My understanding is that with `from __future__ import annotations`, you don't need `typing.Dict`, you can just use `dict`. So this change is unnecessary, I think.

File: src/allmydata/test/cli/wormholetesting.py [hunk 2]
```diff
@@ -61,23 +60,23 @@ class MemoryWormholeServer(object):
...
-        journal=None,
-        tor=None,
-        timing=None,
-        stderr=stderr,
-        _eventual_queue=None,
-        _enable_dilate=False,
-    ):
+        appid: Any,
+        relay_url: Any,
+        reactor: Any,
```
Comment (itamarst) [inline src/allmydata/test/cli/wormholetesting.py:70]: For reactor you can use one of the interfaces in `twisted.internet.interfaces`, there's a variety (not sure if there's a catch-all interface, there might be.)

**Review by itamarst — CHANGES_REQUESTED** [review]
Would be good to have passing tests.

**Comment by itamarst** [comment]
Thank you for working on this! Having good type annotation coverage would make life easier.

File: src/allmydata/test/cli/wormholetesting.py [hunk 3]
```diff
@@ -1,29 +1,22 @@
 """
 An in-memory implementation of some of the magic-wormhole interfaces for
 use by automated tests.
-
```
Comment (hacklschorsch) [inline src/allmydata/test/cli/wormholetesting.py:4]: Why delete all this white space lines?  They seem to be there on purpose.

File: src/allmydata/test/cli/wormholetesting.py [hunk 4]
```diff
@@ -105,9 +95,8 @@ def _view(self, key: ApplicationKey) -> _WormholeServerView:
 @frozen
 class TestingHelper(object):
     """
-    Provide extra functionality for interacting with an in-memory wormhole
+    Provide extra functionality for interacting with d in-memory wormhole
```
Comment (hacklschorsch) [inline src/allmydata/test/cli/wormholetesting.py:98]: Typo?

File: mypy.ini [hunk 5]
```diff
@@ -1,3 +1,6 @@
 [mypy]
 ignore_missing_imports = True
 plugins=mypy_zope:plugin
+show_column_numbers = True
+pretty = True
+show_error_codes = True
```
Comment (exarkun) [inline mypy.ini:6]: Please adjust your editor to add trailing newlines.

**Comment by exarkun** [comment]
The Nix job failures are a result of `python setup.py update_version` failing to write out a version number in the expected place - apparently because there are no release tags in the fork so it can't make up its mind about what the version should be.

File: src/allmydata/test/cli/wormholetesting.py (hunk 1 above)
Comment (danielzwlee) [reply src/allmydata/test/cli/wormholetesting.py]: Changed back to dict. Thank you for your feedback!

File: src/allmydata/test/cli/wormholetesting.py (hunk 2 above)
Comment (danielzwlee) [reply src/allmydata/test/cli/wormholetesting.py:70]: Thanks for pointing me in the right direction. Looked into it there does not seem like there is an all encompassing reactor interface.

File: src/allmydata/test/cli/wormholetesting.py (hunk 4 above)
Comment (danielzwlee) [reply src/allmydata/test/cli/wormholetesting.py]: Fixed.

**Comment by danielzwlee** [comment]
Passes all tests on https://app.circleci.com/pipelines/github/danielzwlee/tahoe-lafs/32/workflows/4518b594-f410-45ed-a734-41d71f7c6035 and failing test is similar to master.

File: mypy.ini [hunk 6]
```diff
@@ -5,6 +5,5 @@ show_column_numbers = True
 pretty = True
 show_error_codes = True
 warn_unused_configs =True
-no_implicit_optional = True
```
Comment (exarkun) [inline mypy.ini:8]: This seems like a merge mistake?

File: src/allmydata/web/status.py [hunk 7]
```diff
@@ -550,7 +550,7 @@ def events(self, req, tag):
             length = r_ev["length"]
             bytes_returned = r_ev["bytes_returned"]
             decrypt_time = ""
-            if bytes:
+            if isinstance(r_ev, bytes):
                 decrypt_time = self._rate_and_time(bytes_returned, r_ev["decrypt_time"])
```
[the remark this answers is not in the document]
Comment (danielzwlee) [reply src/allmydata/web/status.py:554]: Fixed in another PR.

File: mypy.ini (hunk 6 above)
Comment (exarkun) [reply mypy.ini:8]: bump

File: src/allmydata/util/_eliot_updates.py [hunk 8]
```diff
@@ -57,14 +59,14 @@ class _CustomEncoderMemoryLogger(_MemoryLogger):
     This is only necessary on Python 2 where we use an old version of Eliot
     that does not parameterize the encoder.
     """
-    def __init__(self, encoder=eliot_json_encoder):
+    def __init__(self, encoder=eliot_json_encoder)-> None:
```
[the remark this answers is not in the document]
Comment (exarkun) [reply src/allmydata/util/_eliot_updates.py:62]: bump

File: src/allmydata/util/jsonbytes.py [hunk 9]
```diff
@@ -30,8 +30,8 @@ def backslashreplace_py2(ex):
 
 def bytes_to_unicode(any_bytes, obj):
     """Convert bytes to unicode.
-
     :param any_bytes: If True, also support non-UTF-8-encoded bytes.
+
```
[the remark this answers is not in the document]
Comment (exarkun) [reply src/allmydata/util/jsonbytes.py:34]: bump

File: src/allmydata/test/cli/wormholetesting.py [hunk 10]
```diff
@@ -134,18 +133,25 @@ async def wait_for_wormhole(self, appid: AppId, relay_url: RelayURL) -> IWormhol
...
 
 
-def _verify():
+def _verify()-> None:
     """
     Roughly confirm that the in-memory wormhole creation function matches the
     interface of the real implementation.
     """
     # Poor man's interface verification.
+    # Upgrade to getfullargspec since getargspec deprecated since python 3.5
```
[the remark this answers is not in the document]
Comment (exarkun) [reply src/allmydata/test/cli/wormholetesting.py:142]: bump

File: src/allmydata/test/cli/wormholetesting.py [hunk 11]
```diff
@@ -262,7 +268,7 @@ def when_code(self) -> Deferred[WormholeCode]:
             return d
         return succeed(self._code)
 
-    def get_welcome(self):
+    def get_welcome(self)->Deferred[str]:
```
[the remark this answers is not in the document]
Comment (exarkun) [reply src/allmydata/test/cli/wormholetesting.py:271]: bump

File: src/allmydata/test/cli/wormholetesting.py [hunk 12]
```diff
@@ -276,8 +282,8 @@ def when_received(self) -> Deferred[WormholeMessage]:
             )
         d = self._view.wormhole_by_code(self._code, exclude=self)
 
-        def got_wormhole(wormhole):
-            msg = wormhole._payload.get()
+        def got_wormhole(wormhole: _MemoryWormhole)-> Deferred[Any]:
```
[the remark this answers is not in the document]
Comment (exarkun) [reply src/allmydata/test/cli/wormholetesting.py:285]: If we know the type of the payload then we don't need the `Any` in this `Deferred[Any]` or the one on the next line.

**Comment by danielzwlee** [comment]
Closed to open a new clean target branch for merging.

**Pull request closed without merging**